### ISSUE TITLE
float16 gradients are nan/inf in rotation_matrix_to_quaternion, rotation_matrix_to_axis_angle and normalize_quaternion: the eps guards flush to zero in half

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4623]
`rotation_matrix_to_quaternion`, `rotation_matrix_to_axis_angle` and `normalize_quaternion` return `nan`/`inf` gradients in float16, because their `eps` guards flush to zero in that dtype.

Same family as #4229, and the same shape as the fix in #4613 (`axis_angle_to_rotation_matrix`). Filing so the remaining sites are tracked rather than fixed piecemeal.

## Reproduction

```python
import torch
from kornia.geometry.conversions import (
    rotation_matrix_to_quaternion, rotation_matrix_to_axis_angle, normalize_quaternion,
)

for fn, x in ((rotation_matrix_to_quaternion, torch.eye(3)),
              (rotation_matrix_to_axis_angle,  torch.eye(3)),
              (normalize_quaternion,           torch.zeros(4))):
    t = x.half().requires_grad_(True)
    fn(t).sum().backward()
    print(fn.__name__, t.grad.flatten()[:4].tolist())
```

On `ecbcad561`, torch 2.14.0, CPU:

```
rotation_matrix_to_quaternion  [nan, -0.25, 0.25, 0.25]
rotation_matrix_to_axis_angle  [nan, -0.5,  0.5,  0.5]
normalize_quaternion           [inf, inf,   inf,  inf]
```

## Scope

| site | float16 | bfloat16 | float32 | float64 |
| --- | --- | --- | --- | --- |
| `rotation_matrix_to_quaternion(eye(3))` | **non-finite** | finite | finite | finite |
| `rotation_matrix_to_quaternion(diag(1,-1,-1))` — a 180° rotation | **non-finite** | finite | finite | finite |
| `rotation_matrix_to_axis_angle(eye(3))` | **non-finite** | finite | finite | finite |
| `normalize_quaternion(zeros(4))` | **non-finite** | finite | finite | finite |

`rotation_matrix_to_axis_angle` is `rotation_matrix_to_quaternion` → `quaternion_to_axis_angle`, so it inherits the first row rather than being an independent defect.

## Mechanism

Two distinct causes, both "the guard is unrepresentable in the dtype".

**1. `rotation_matrix_to_quaternion` — `sqrt` at zero in a discarded `torch.where` arm.** The four branches are computed for every element and selected afterwards. At `eye(3)`, `cond_1`'s radicand `1 + m00 - m11 - m22` is exactly `0`, and the branch computes `sqrt(radicand + eps)` with `eps = 1e-8`:

```
torch.float16:  1e-8 as float16 = 0.0     -> d/dr sqrt(r+eps) at r=0 : inf
torch.float32:  1e-8 as float32 = 9.99999993922529e-09  -> d/dr sqrt(r+eps) at r=0 : 5000.0
```

`inf` × the `where`'s zero upstream gradient = `nan`. In float32 the same expression gives a large but finite 5000, which is why only float16 is affected. bfloat16 keeps `1e-8` (`1.0011717677116394e-08`), so it is finite there too.

This is precisely the trap AGENTS.md documents: a floor below the dtype's smallest representable value is not a guard, and `torch.where` does not stop the discarded arm from being differentiated.

**2. `normalize_quaternion` — the reciprocal of the floor overflows float16.** `eps` defaults to `1e-12` and is floored at the dtype's smallest positive subnormal. In float16 that is `5.96e-08`, and `1/5.96e-08 ≈ 1.7e7`, well past float16's max of `65504` → `inf`. float32 gives the finite `999999995904.0`.

## Suggested fix

The same substitution #4613 uses, and the one AGENTS.md prescribes: put a safe argument into the expression that gets differentiated, rather than flooring the unsafe one.

```python
mask = radicand > threshold
safe = torch.where(mask, radicand, torch.ones_like(radicand))
out  = torch.where(mask, safe.sqrt(), other_branch)
```

`quaternion_to_axis_angle`, `axis_angle_to_quaternion`, `quaternion_to_rotation_matrix`, `So3.exp` and `Se3.exp` already use this pattern and are clean — so there is an in-tree reference for the shape of the fix.

## Interaction with open work

- **#4613** fixes `axis_angle_to_rotation_matrix` with exactly this substitution. Its approach is the model.
- **#3981** touches `rotation_matrix_to_quaternion` and would currently make this *worse*, not better: it replaces `sqrt(x + eps)` with `sqrt(x.clamp(min=eps))`, and its own docstring says "`1e-8` flushes to `0.0` in `float16`". Per AGENTS.md, `clamp` is not a portable gradient guard — at the bound its derivative is `1.0` on torch 2.5.1/2.9.1 and `0.0` on 2.14.0, and where the bound is also `sqrt`'s singular point the older half of the supported range still leaks `inf`. That PR's outstanding review ask is already "version-independent gradient guard, half coverage"; this issue is the measurement behind it.
- #3951 and #3952 are about **forward** values (non-unit quaternions), not gradients, so they are separate.

## Notes

- Measured on `ecbcad561`, torch 2.14.0, macOS arm64, CPU.
- Any fix needs a half-precision gradient pin that actually discriminates. Per #4613's review, a finiteness-only assertion is not enough — `BaseTester`'s float16 tolerance is `atol=1e-3`, so an `isfinite` check plus an O(1e-4) value assert can be satisfied by a badly wrong gradient. Pin a value against a float64 reference with explicit tolerances.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by PushpakAg** [comment]
Seems like the `eps` value in float16 isn't handling small values properly, leading to `nan` or `inf` gradients. I'd check how `eps` is being used in those functions when dealing with float16.

**Comment by NiteeshGowda72** [comment]
Could we avoid `torch.clamp(..., min=safe_eps)` in `normalize_quaternion` as well?

The issue description specifically calls out `.clamp(min=eps)` as a non-portable gradient guard, and the PR otherwise follows the safe-argument substitution pattern from #4613.

For the zero-norm case, `safe_norm = torch.where(mask, norm, torch.ones_like(norm))` already gives us a safe denominator. Would it be possible to structure the nonzero path without the clamp, so the denominator passed to the division is always the safe substituted value?

I think this would make the `normalize_quaternion` fix fully consistent with the numerical-stability pattern described in #4623.

**Comment by PushpakAg** [comment]
I'll look into whether we can structure the nonzero path to avoid using `torch.clamp` in `normalize_quaternion` and ensure the denominator is always a safe substituted value. That should align with the numerical stability pattern mentioned.

**Comment by ducha-aiki** [comment]
Fixed on `main` by `96671e014` (#3981, which carried #4627's commits). The finite float16 gradients of `rotation_matrix_to_quaternion`, `rotation_matrix_to_axis_angle` and `normalize_quaternion` are pinned by the `*_4623` tests in `tests/geometry/test_conversions.py`, which pass in float16, float32 and float64.

The squash message cites this issue only as "(#4623)", with no closing keyword, so I'm closing it by hand.

Posted on behalf of @ducha-aiki by Claude (Opus 5).
